src/TraceEvent/EventPipe/RewindableStream.cs

```diff
@@ -0,0 +1,167 @@
+            if (bytesRead > 0)
+            {
+                _cachedBytes.AsSpan(_cacheArrayOffset, bytesRead).CopyTo(buffer);
+                _cacheArrayOffset += bytesRead;
+                _cacheStreamOffset += bytesRead;
+                _cacheBytesAvailable -= bytesRead;
+            }
+            return bytesRead;
+        }
+
+        public T Read<T>() where T : struct
+        {
+            Span<byte> buffer = stackalloc byte[Unsafe.SizeOf<T>()];
+            Read(buffer);
+            return MemoryMarshal.Read<T>(buffer);
+        }
+
+        public void Read(Span<byte> buffer) => ReadAtLeast(buffer, buffer.Length);
+
+        public int ReadAtLeast(Span<byte> buffer, int minLength)
+        {
+            // consume bytes from the cache first
+            int bytesRead = ReadFromCache(buffer);
+
+            // if the cache didn't satisfy the minimum read, read from the stream
+            while (bytesRead < minLength)
+            {
+                Debug.Assert(_cacheBytesAvailable == 0);
+                int read = _stream.Read(buffer.Slice(bytesRead));
+                if (read == 0)
+                {
+                    throw new EndOfStreamException();
+                }
+                bytesRead += read;
+                _cacheStreamOffset += read;
+            }
+            return bytesRead;
+        }
+
+        // Readers are allowed to read more than they need, then give the extra bytes back to be stored in the cache for later.
+        // We only support caching bytes from the most recent call to Read/ReadAtLeast, and only up to the maximum cache size
+        // specified in the constructor.
+        public void CacheUnusedBytes(ReadOnlySpan<byte> extraBytes)
+        {
+            if (extraBytes.Length + _cacheBytesAvailable > MaxCacheSize)
+            {
+                // This should never happen unless there is a bug in the code
+                throw new InvalidOperationException($"Cache overflow");
+            }
+            // copy the new data into the cache
+            _cacheBytesAvailable += extraBytes.Length;
+            _cacheStreamOffset -= extraBytes.Length;
+            _cacheArrayOffset -= extraBytes.Length;
+            extraBytes.CopyTo(_cachedBytes.AsSpan(_cacheArrayOffset));
+        }
+
+
+        public void Dispose()
+        {
+            _stream.Dispose();
+        }
+    }
+
+    internal static class StreamExtensions
+    {
+        delegate int StreamReadDelegate(Stream stream, Span<byte> buffer);
+
+        private static readonly StreamReadDelegate s_streamReadDelegate = InitSpanRead();
+
+        private static StreamReadDelegate InitSpanRead()
+        {
+            // Opportunistically use it if reflection shows it is available, otherwise fallback to
+            // the slower Read(byte[], int, int) method + a copy.
+            MethodInfo method = typeof(Stream).GetMethod("Read", new Type[] { typeof(Span<byte>) });
```

Comment:
It would be good to have a test that ensures that this works properly for platforms that have the `Read(Span<byte>)` method.  TraceEventTests targets and runs on both .NET Framework and .NET 8, so we should be able to cover both.

Reply:
Sure, I should be able to do something like that.